add 2captcha

xMordax opened the pull request:
I added 2captcha to solve the issue with immobilienscout. To solve that I use Selenium and 2captcha now, it is optional to use it.

codders commented:
Thanks for this - looks great. Just a couple of comments...

codders commented on flathunter/config.py:
```diff
@@ -9,14 +9,11 @@
 from flathunter.crawl_immowelt import CrawlImmowelt
 from flathunter.filter import Filter
 
+
```
Looks like some random whitespace - please remove it.

codders commented on flathunter/config.py:
```diff
@@ -27,6 +24,10 @@ def __init__(self, filename=None, string=None):
         self.__log__.info("Using config %s", filename)
         with open(filename) as file:
             self.config = yaml.safe_load(file)
+        self.__searchers__ = [CrawlImmobilienscout(self.config),
```
In the case where the config is supplied as a string, this initialization fails the constructor returns before self.__searchers__ gets set. Need to replace the 'return' on line 21 with an 'else' for the file case so that both paths initialise self.__searchers__.

Also, here you pass self.config (which is the internal dictionary) rather than self (the config object). It's cleaner if the external API to the configuration is via the config object - I would like eventually for code paths not to directly access config keys.

codders commented on flathunter/crawl_immobilienscout.py:
```diff
@@ -5,15 +5,35 @@
 from flathunter.abstract_crawler import Crawler
 
+
 class CrawlImmobilienscout(Crawler):
```
Again whitespace here.

codders commented on flathunter/crawl_immobilienscout.py:
```diff
@@ -5,15 +5,35 @@
     RESULT_LIMIT = 50
 
-    def __init__(self):
+    def __init__(self, config):
```
I would prefer this argument to be the config object.

codders commented on flathunter/crawl_immobilienscout.py:
```diff
@@ -5,15 +5,35 @@
+        self.captcha_api_key = None
+        self.checkbox = None
+        self.afterlogin_string = None
+        if "captcha" in config:
```
This could be replaced with a function call - config.captcha_enabled() - which does the same check internally in the config object.

codders commented on flathunter/crawl_immobilienscout.py:
```diff
@@ -5,15 +5,35 @@
+        self.checkbox = None
+        self.afterlogin_string = None
+        if "captcha" in config:
+            self.captcha_api_key = self.config.get('captcha', dict()).get('api_key', '')
```
All these lines have the same `self.config.get('captcha', dict())` construct. I think it makes sense to use a local variable here for that value.

codders reviewed (commented):
I'm really super grateful that you went to the effort to write this. I tested it on my machine and it works great. I have some feedback, but I don't want to make your life difficult - I've already implemented the feedback changes on a branch in my repo based on your PR: https://github.com/codders/flathunter/tree/feature/2captcha . If you want you can copy from there. Important for me is that the test suite passes (in your branch, test/test_processor.py is not passing for the reasons mentioned in the review). The rest is just stylistic and I'm not so committed to it.

xMordax commented:
> I'm really super grateful that you went to the effort to write this. I tested it on my machine and it works great. I have some feedback, but I don't want to make your life difficult - I've already implemented the feedback changes on a branch in my repo based on your PR: https://github.com/codders/flathunter/tree/feature/2captcha . If you want you can copy from there. Important for me is that the test suite passes (in your branch, test/test_processor.py is not passing for the reasons mentioned in the review). The rest is just stylistic and I'm not so committed to it.

I completely forgot about the tests cases, thanks for taking the time and doing the improvements, I added it to the PR. I also updated bit the README and added a comment to the code.

codders reviewed (approved):
lgtm - thanks!
